### PR TITLE
Modular Taint Analysis Improvements

### PR DESCRIPTION
**Pull request by ArquintL**
- explicit sink & source types
- improves some error messages & warnings incl. better positional information
- fixes check whether a method or loop terminates

### CODE REVIEW

File: src/main/scala/viper/silver/plugin/standard/reasoning/analysis/VarAnalysisGraphMap.scala [hunk 1]
```diff
@@ -307,94 +405,102 @@ case class VarAnalysisGraphMap(prog: Program,
...
-        val heapDependent = Expressions.couldBeHeapDependent(f.func(prog), prog)
-        f.args.flatMap(e => getLocalVarDeclsFromExpr(e).filter(v => v.typ != Ref)).toSet ++ (if (heapDependent) Set(heapVertex) else Set())
-      case DomainFuncApp(_, exps, _) => exps.flatMap(e => getLocalVarDeclsFromExpr(e)).toSet
-      case _: ForPerm | _: CurrentPerm => Set(heapVertex)
-      case FieldAccess(v, _) => getLocalVarDeclsFromExpr(v) + heapVertex
-      case AccessPredicate(access, _) => getLocalVarDeclsFromExpr(access) + heapVertex
-      case _ => Set()
+        val isHeapDependent = Expressions.couldBeHeapDependent(f.func(prog), prog)
+        // TODO: do we need to filter out Refs?
+        f.args.flatMap(e => getSinksFromExpr(e)).toSet ++ (if (isHeapDependent) Set(HeapSink) else Set.empty)
```
Comment (ArquintL) [inline src/main/scala/viper/silver/plugin/standard/reasoning/analysis/VarAnalysisGraphMap.scala:507]: Here I was unsure about porting the existing code. The `Ref` filtering imho seems to get rid of the heap node but wouldn't this get added anyway by taking the union if the singleton set containing the heap? I'm literally missing the intuition behind the existing code as it seems unclear why it's okay to filter out `Ref`

**Comment by ArquintL** [comment]
Don't worry about the conflict for now, I'll fix them later today or tomorrow

**Comment by ArquintL** [comment]
Merge conflicts are resolved

File: src/main/scala/viper/silver/plugin/standard/reasoning/analysis/VarAnalysisGraphMap.scala (hunk 1 above)
Comment (jogasser) [reply src/main/scala/viper/silver/plugin/standard/reasoning/analysis/VarAnalysisGraphMap.scala:507]: You are right, I did not question this when I first saw it. I think your solution is right, as this example should not be allowed imo but would pass with the previous approach:
```
function test(e: Ref): Ref {
    e
}

method testM() {
    var x: Ref
    prove forall i:Ref assuming true implies i == x {
        x := test(i)
    }
}
```

Comment (ArquintL) [reply src/main/scala/viper/silver/plugin/standard/reasoning/analysis/VarAnalysisGraphMap.scala:507]: I agree. Please add this as a testcase

**Pull request merged**